### PR TITLE
MBS-13562: Check editing unlimited strings

### PR DESCRIPTION
**Pull request by yvanzo**
# MBS-13562

See the commit messages for details.

The user interface has not been much updated for user-friendliness as this is an edge case.

# Testing

Added automated:
* [x] unit tests for each validation subroutine
* [x] integration tests for each check implementation part (`Role::CheckOverlongString`, `Role::EditArtistCredit`, `Medium::Util`)

I didn’t add automated Selenium tests as there is no change to the user interface beyond the explanatory message.

Manually tested editing a sample database with:
* [x] creating/editing an artist with overlong name/sort name
* [x] creating/editing a recording with overlong name/artist credit join phrase
* [x] creating/editing a release with overlong name/artist credit join phrase
* [x] creating/editing a release with overlong track name/track artist credit join phrase

# Documenting

* [ ] Mention this edge case in the [Editing_FAQ page](https://wiki.musicbrainz.org/Editing_FAQ)
* [ ] Update the ticket according to the actual implementation.

# Further action

1. Deploy in production
2. Merge to master
3. Release beta
4. Make the updated Editing_FAQ page visible on MusicBrainz website

### CODE REVIEW

File: lib/MusicBrainz/Server/Validation.pm [hunk 1]
```diff
@@ -98,6 +102,12 @@ sub is_database_bigint_id {
     is_positive_integer($t) and $t <= $MAX_POSTGRES_BIGINT;
 }
 
+sub has_oneline_string_length {
```
Comment (reosarevok) [inline lib/MusicBrainz/Server/Validation.pm:113]: This name is confusing to me - seeing it I wouldn't know if it means it's over the limit or under the limit. What about something boring like `is_not_too_long_string`?

File: lib/MusicBrainz/Server/Constants.pm [hunk 2]
```diff
@@ -426,6 +427,8 @@ Readonly our $MAX_INITIAL_TRACKS => 100;
 Readonly our $MAX_POSTGRES_INT => 2147483647;
 Readonly our $MAX_POSTGRES_BIGINT => 9223372036854775807;
 
+Readonly our $MAX_ONELINE_STRING_LENGTH => 1024;
```
Comment (reosarevok) [inline lib/MusicBrainz/Server/Constants.pm:430]: Is "oneline" a technical term? Is there a reason why you wouldn't just call it `$MAX_POSTGRES_INDEX_VARCHAR` or something to mirror the `INT` and `BIGINT` constants above?

File: lib/MusicBrainz/Server/Validation.pm [hunk 3]
```diff
@@ -102,6 +103,13 @@ sub is_database_bigint_id {
     is_positive_integer($t) and $t <= $MAX_POSTGRES_BIGINT;
 }
 
+sub is_database_indexable_string {
```
Comment (reosarevok) [inline lib/MusicBrainz/Server/Validation.pm:106]: Why do we need both of this and the 1024 limit? Wouldn't this check be enough for our needs? Or is it just to have a more easily communicable limit?

File: lib/MusicBrainz/Server/Edit/Role/EditArtistCredit.pm [hunk 4]
```diff
@@ -22,7 +26,21 @@ around initialize => sub {
...
+        } else {
+            my $new_artist_credit_full_name = '';
+            for my $part (@{ $new_artist_credit->{names} }) {
+              $new_artist_credit_full_name .= $part->{name} . $part->{join_phrase};
+            };
+
+            if (defined($new_artist_credit_full_name) && !(
+                has_oneline_string_length($new_artist_credit_full_name) &&
+                is_database_indexable_string($new_artist_credit_full_name)
+            )) {
```
Comment (reosarevok) [inline lib/MusicBrainz/Server/Edit/Role/EditArtistCredit.pm:40]: We do this check for three conditions often enough that it feels it should be its own function.

File: root/layout.tt [hunk 5]
```diff
@@ -91,6 +91,13 @@
         </div>
         [% END %]
 
+        [% IF overlong_string %]
+        <div class="banner warning-header">
+            <p>[% l('Some text you entered is overlong, please fix the text or
+                     shorten it and use annotation for the full text instead.') %]</p>
```
Comment (reosarevok) [inline root/layout.tt:97]: "overlong" is not a common word. What about "Some text you've entered is too long for us to support. Please shorten it, and if necessary enter the full text in the annotation for reference." or something like that?

File: root/layout/index.js [hunk 6]
```diff
@@ -294,6 +294,17 @@ const Layout = ({
           </div>
         ) : null}
 
+        {$c.stash.overlong_string /*:: === true */ ? (
+          <div className="banner warning-header">
+            <p>
+              {l(
+                `Some text you entered is overlong, please fix the text or
+                 shorten it and use annotation for the full text instead.`,
+              )}
```
Comment (reosarevok) [inline root/layout/index.js:303]: Same as above.

File: lib/MusicBrainz/Server/Validation.pm (hunk 1 above)
Comment (yvanzo) [reply lib/MusicBrainz/Server/Validation.pm:113]: It would be less confusing w.r.t. “over/under” but it doesn’t say for what it “is not too long”.

Would `is_short_oneline_string` carry both meanings?

File: lib/MusicBrainz/Server/Constants.pm (hunk 2 above)
Comment (yvanzo) [reply lib/MusicBrainz/Server/Constants.pm:430]: > Is "oneline" a technical term?

It is opposed to multiple-lines strings such as annotations which are of variable unlimited length too in Postgres.

> Is there a reason why you wouldn't just call it `$MAX_POSTGRES_INDEX_VARCHAR` or something to mirror the `INT` and `BIGINT` constants above?

It can also be a `TEXT` in Postgres and this is more our decision than a Postgres limitation really.

File: lib/MusicBrainz/Server/Edit/Role/EditArtistCredit.pm (hunk 4 above)
Comment (yvanzo) [reply lib/MusicBrainz/Server/Edit/Role/EditArtistCredit.pm:40]: Would a function for all of these conditions in our `Validation` module do?

File: root/layout.tt (hunk 5 above)
Comment (yvanzo) [reply root/layout.tt]: > "overlong" is not a common word. What about "Some text you've entered is too long for us to support.

It is not a common word but it isn’t overcomplicated to understand even the first time you read it. The fact that it is not common makes it easier to search the documentation for as we can refer to it in the Editing FAQ.

> Please shorten it, and if necessary enter the full text in the annotation for reference." or something like that?

Committed through https://github.com/metabrainz/musicbrainz-server/pull/3251/commits/ef4fb7e32e76b0ae2651d67becda87062710cf8d.

File: lib/MusicBrainz/Server/Edit/Role/EditArtistCredit.pm (hunk 4 above)
Comment (reosarevok) [reply lib/MusicBrainz/Server/Edit/Role/EditArtistCredit.pm:40]: Sure :)

File: lib/MusicBrainz/Server/Validation.pm (hunk 1 above)
Comment (reosarevok) [reply lib/MusicBrainz/Server/Validation.pm:113]: `is_short_enough_oneline_string`? `is_short_enough_single_line_string`? Just `is_short` seems a bit funny because 1024 characters is not short :D But I can live with that too really.

Comment (yvanzo) [reply lib/MusicBrainz/Server/Validation.pm]: I renamed it `has_at_most_oneline_string_length` in https://github.com/metabrainz/musicbrainz-server/pull/3251/commits/4ee30f845aa06df6482640b4c92fa0743aca2641 to keep it meaningful and not confusing.

File: lib/MusicBrainz/Server/Edit/Role/EditArtistCredit.pm (hunk 4 above)
Comment (yvanzo) [reply lib/MusicBrainz/Server/Edit/Role/EditArtistCredit.pm]: I added `is_overlong_string` in https://github.com/metabrainz/musicbrainz-server/pull/3251/commits/abe06ef84d8655caffbdff15da2924593bbcb412.

File: lib/MusicBrainz/Server/Validation.pm (hunk 3 above)
Comment (yvanzo) [reply lib/MusicBrainz/Server/Validation.pm:107]: > I did check those, but I'm still a bit confused. It seems the real limit that triggered the bug was actually the 2704 bytes one, no?

The bug MBS-13555, yes. And it would also prevent MBS-13536 to happen again.

> Wouldn't it be enough then to check for 2704 bytes and not worry about the amount of characters and whether they are multibyte or not?

As explained in commit message, it is more reasonable to additionally limit to 1024 characters.

Comment (mwiencek) [reply lib/MusicBrainz/Server/Validation.pm:107]: IIUC, the 8192 limit from MBS-13536 is for all indexes generally, and the 2704 limit is for btree indexes. The index from MBS-13536 was *also* a btree index, but perhaps it hit the "general" limit first before it checked the more specifiec btree limits (but I am just guessing).

> As explained in commit message, it is more reasonable to additionally limit to 1024 characters.

I agree that a fixed character limit is reasonable and also easier to explain to users -- though I'm not sure if the majority of the offenders are mistakes or intended to cause harm:

1055 characters - https://musicbrainz.org/edit/111225345
1056 characters - https://musicbrainz.org/edit/111225330
1126 characters - https://musicbrainz.org/edit/111225347
1176 characters - https://musicbrainz.org/edit/111225344
1377 characters - https://musicbrainz.org/edit/111225342

I'm not familiar with any of the source material though.

Given the additional btree-indexable byte limit check, we could perhaps be a little more generous and use a less-geeky number like 1500 which would cover most of the existing cases. Or maybe allow auto-editors to bypass the limit (as long as it's under 2704 bytes still).

Comment (yvanzo) [reply lib/MusicBrainz/Server/Validation.pm:107]: These are all examples of mistakes:

> 1055 characters - https://musicbrainz.org/edit/111225345

This is a clearly list of works that should have been written in the liner notes and would better fit in work relationships, or in annotation to start with.

It should be: “Musicalische Exequien, SWV 279, Concert in Form einer teutschen Begräbnis-Missa: II”

> 1056 characters - https://musicbrainz.org/edit/111225330

Same: “Transmission 7”

> 1126 characters - https://musicbrainz.org/edit/111225347

Same: “24 Duos Op.186 (For Two Saxophones)”

> 1176 characters - https://musicbrainz.org/edit/111225344

Same: “Korean Cosmetics”

No work relationship here though. This is actually a recipe, it would just better fit in annotation. The above name of the recipe is even missing from the current data if we compare to the linked [Bandcamp page](https://absenterratum.bandcamp.com/track/ingredients-water-eau-methylpropanediol-glycerin-ethyl-perfluoroisobutyl-ether-ethyl-perfluorobutyl-ether-sapindus-mukorossi-fruit-extract-acrylates-copolymer-betaine-hydroxyacetophenone-polyacry). Just using this name as track title would make the tracklist much more readable.

> 1377 characters - https://musicbrainz.org/edit/111225342

Same: “New Numbers”

Comment (mwiencek) [reply lib/MusicBrainz/Server/Validation.pm:107]: Thanks for checking them; it'd be nice to replace your existing edits with the suggestions you're certain about when you have time, since it's much nicer than truncation, and it may be unlikely for some of these releases to be fixed for a long time! (Note I didn't link to any of the "creative writing" ones.)

File: lib/MusicBrainz/Server/Constants.pm (hunk 2 above)
Comment (mwiencek) [reply lib/MusicBrainz/Server/Constants.pm:430]: Maybe worth mentioning as a comment in the source code (outside of the commit message) how the constant was derived? (i.e., the "maximum size allowed for the index type" divided by four to account for multibyte characters).

File: t/lib/t/MusicBrainz/Server/Edit/Medium/Edit.pm [hunk 7]
```diff
@@ -710,6 +711,74 @@ test 'Tracklist merging (MBS-8752 / MBS-7475)' => sub {
...
+        artist_credit => 1,
+    });
+
+    try {
+        like exception {
+            create_edit($c, $medium, [
+                Track->new(
+                    artist_credit => $c->model('ArtistCredit')->get_by_id(1),
+                    is_data_track => 0,
+                    name => ('𝄞𝄵𝅘𝅥𝅘𝅥𝅮𝅘𝅥𝅮𝅘𝅥𝅮𝄾𝅘𝅥𝄀𝅘𝅥𝅮𝅘𝅥𝅮𝅘𝅥𝅮𝄾𝆑𝅗𝅥𝄂' x 64),
```
Comment (mwiencek) [inline t/lib/t/MusicBrainz/Server/Edit/Medium/Edit.pm:732]: A minor suggestion here would be to add `no utf8;` to the enclosing block in case we ever add `use utf8;` to the file for some reason and accidentally break this test.

File: lib/MusicBrainz/Server/Validation.pm (hunk 3 above)
Comment (reosarevok) [reply lib/MusicBrainz/Server/Validation.pm:107]: The 24 Duos one is a wrong Discogs import where a lot of tracks got smashed together, I'll fix that one. I disagree that Transmission 7 is wrong, but I think it'd be ok either way - if you change that one you should change all though. "Korean Cosmetics" seems to be the artist name and should probably not be added to the title - I'd just find a good place to put `...` at the end of that one, maybe just do `Moringa Pterygosperma Seed Extract...`? New Numbers seems fine to change to just that. The Musicalische Exequien, SWV 279 is also not wrong, but shortening it is probably sensible. That disc has other errors, so I'll do that part myself as well.

Comment (yvanzo) [reply lib/MusicBrainz/Server/Validation.pm:107]: Sorry but no, the changes should probably be made for the whole releases not just these tracks, this isn’t worth my time, and there is a lot of trash stuff around (such as the above second example with 1056 characters) that I don’t even want to read, that’s why I made a script to truncate names arbitrarily rather than editing by hand. I’m not a classical editor either, so the classical releases would be better edited by some more knowledgeable editor than me about classical releases. I’m sure that some interested editors are following these edits that are linked from the [ticket’s comment](https://tickets.metabrainz.org/browse/MBS-13562?focusedId=72046&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-72046) and will make the most appropriate changes in time (likely after this check is deployed). As I wrote on IRC before, the data isn’t completely lost since it lies in the editing history anyway. There should be nothing wrong with truncating a few names arbitrarily for the sake of maintaining the database. If we decide to relax this constraint later on, it will just take changing a constant in code and reverting a few edits.

Comment (mwiencek) [reply lib/MusicBrainz/Server/Validation.pm:107]: Thanks @reosarevok -- I guess we're all okay with the existing checks then. I didn't mean to start bikeshedding on a specific number, but was just hesitant to restrict possibly-valid data. @yvanzo, it sounds like you can just cancel your existing edits if reosarevok is looking into them.

File: t/lib/t/MusicBrainz/Server/Edit/Medium/Edit.pm (hunk 7 above)
Comment (yvanzo) [reply t/lib/t/MusicBrainz/Server/Edit/Medium/Edit.pm:732]: I’m not sure to follow here because the `Validation` tests have the same test string and `use utf8;` already.

Comment (mwiencek) [reply t/lib/t/MusicBrainz/Server/Edit/Medium/Edit.pm:732]: Sorry, I gave wrong info. If these are utf8 strings with multi-byte characters, the files should have `use utf8;`. Otherwise I don't think `length` will work properly when they are passed elsewhere...

```Perl
# P1.pm
package P1;

#use utf8;
our $string = '𝄞𝄵𝅘𝅥𝅘𝅥𝅮𝅘𝅥𝅮𝅘𝅥𝅮𝄾𝅘𝅥𝄀𝅘𝅥𝅮𝅘𝅥𝅮𝅘𝅥𝅮𝄾𝆑𝅗𝅥𝄂' x 64;
```

```Perl
# P2.pm
package P2;

use P1;

print length($P1::string) . "\n";
use bytes;
print length($P1::string) . "\n";
```

With `#use utf8;` commented out it will print `4096` for the first length check, but with `use utf8;` enabled it will print `1024`.

File: lib/MusicBrainz/Server/Constants.pm (hunk 2 above)
Comment (yvanzo) [reply lib/MusicBrainz/Server/Constants.pm:430]: Added comments for both constants. See https://github.com/metabrainz/musicbrainz-server/compare/182f1255fa7f9ebf63853b353812f5e0d26d4033..50f59153b7f5ca57a5cbf4f943b917307aafbc7a

File: lib/MusicBrainz/Server/Validation.pm (hunk 3 above)
Comment (yvanzo) [reply lib/MusicBrainz/Server/Validation.pm:107]: > Thanks @reosarevok -- I guess we're all okay with the existing checks then. I didn't mean to start bikeshedding on a specific number, but was just hesitant to restrict possibly-valid data.

Initially I was more leaning towards 512 bytes to be honest. I picked up 1024 bytes after your own data analysis on IRC as there is indeed no legitimate use case beyond it. Just reminding that [titles](https://en.wikipedia.org/wiki/Title_(publishing)) are supposed to be “minimal summaries”, neither dissertations nor liner notes.

> @yvanzo, it sounds like you can just cancel your existing edits if reosarevok is looking into them.

It wouldn’t hurt to approve all of these edits really, even if these titles are further trimmed later on by volunteers.

File: t/lib/t/MusicBrainz/Server/Edit/Medium/Edit.pm (hunk 7 above)
Comment (yvanzo) [reply t/lib/t/MusicBrainz/Server/Edit/Medium/Edit.pm:733]: Good catch! Fixed with the commit https://github.com/metabrainz/musicbrainz-server/pull/3251/commits/dfd7f9bca469279bb32ad50d301811a5035b92bd.

File: lib/MusicBrainz/Server/Constants.pm [hunk 8]
```diff
@@ -1073,6 +1077,29 @@ Row ID for the Deleted Artist entity
...
+is generally set to 8192 bytes when compiling Postgres, and also because no
+legitimate examples have been found or are to be expected beyond this length.
+
+The main reason is to bring a reasonable bound for otherwise unlimited titles,
+not only to prevent index issues, but also to make it easier to use for other
+pieces of software accessing this data, and even to improve the data quality.
+
+=item $MAX_POSTGRES_INDEXED_STRING_BYTES
+
+Maximum number of bytes for a string to be indexable by Postgres.
```
Comment (mwiencek) [inline lib/MusicBrainz/Server/Constants.pm:1097]: In case you feel that it's worth mentioning, it's technically specific to btree indexes, but they are the default type of index and we don't use the other types extensively.

**Review by mwiencek — APPROVED** [review]
Thanks for your time on this oversilly issue. ;)

File: lib/MusicBrainz/Server/Constants.pm (hunk 8 above)
Comment (yvanzo) [reply lib/MusicBrainz/Server/Constants.pm:1097]: Totally. Now mentioned. See https://github.com/metabrainz/musicbrainz-server/compare/dfd7f9bca469279bb32ad50d301811a5035b92bd..b0a32248ef6a7e3ccafc18ce771d31bba3e4c5a7
